internal/cli/plugin/update.go

```diff
@@ -0,0 +1,265 @@
+	"github.com/mongodb/mongodb-atlas-cli/atlascli/internal/cli"
+	"github.com/mongodb/mongodb-atlas-cli/atlascli/internal/cli/require"
+	"github.com/mongodb/mongodb-atlas-cli/atlascli/internal/flag"
+	"github.com/mongodb/mongodb-atlas-cli/atlascli/internal/log"
+	"github.com/mongodb/mongodb-atlas-cli/atlascli/internal/plugin"
+	"github.com/mongodb/mongodb-atlas-cli/atlascli/internal/usage"
+	"github.com/spf13/cobra"
+)
+
+var (
+	errTooManyArguments        = errors.New(`either the "--all" flag or the plugin identifier can be provided, but not both`)
+	errNotEnoughArguments      = errors.New(`either the "--all" flag or the plugin identifier needs to be provided`)
+	errPluginHasNoGithubValues = errors.New(`specified plugin does not have any Github values`)
+)
+
+type UpdateOpts struct {
+	cli.OutputOpts
+	Opts
+	UpdateAll bool
+	pluginArg string
+}
+
+func printPluginUpdateWarning(p *plugin.Plugin, err error) {
+	_, _ = log.Warningf("could not update plugin %s because: %v\n", p.Name, err)
+}
+
+// checks if the plugin manifest is valid and that the plugin
+// doesn't contain any commands that conflict with existing CLI commands.
+func (opts *UpdateOpts) validatePlugin(pluginDirectoryPath string) error {
+	// Get the manifest from the plugin directory
+	manifest, err := plugin.GetManifestFromPluginDirectory(pluginDirectoryPath)
+	if err != nil {
+		return err
+	}
+
+	err = validateManifest(manifest)
+	if err != nil {
+		return err
+	}
+
+	// make sure that there is exactly one plugin with the same name
+	pluginCount := 0
+	for _, p := range opts.plugins {
+		if manifest.Name == p.Name {
+			pluginCount++
+		}
+	}
+	if pluginCount != 1 {
+		return fmt.Errorf(`there needs to be exactly 1 plugin with the name "%s", but there are %d`, manifest.Name, pluginCount)
+	}
+
+	// Check for duplicate commands
+	existingCommandsMap := make(map[string]bool)
+	for _, cmd := range opts.existingCommands {
+		// only add command to existing commands map if it is not part of the plugin we want to update
+		if sourcePluginName, ok := cmd.Annotations["sourcePluginName"]; !ok || sourcePluginName != manifest.Name {
+			existingCommandsMap[cmd.Name()] = true
+		}
+	}
+	if plugin.HasDuplicateCommand(manifest, existingCommandsMap) {
+		return fmt.Errorf(`could not load plugin "%s" because it contains a command that already exists in the AtlasCLI or another plugin`, manifest.Name)
+	}
+
+	return nil
+}
+
+func (opts *UpdateOpts) updatePlugin(githubAssetRelease *GithubAssetManager, existingPlugin *plugin.Plugin) error {
+	// get all plugin assets info from github repository
+	assets, err := githubAssetRelease.getPluginAssetInfo()
+	if err != nil {
+		return err
+	}
+
+	// find correct assetID using system requirements
+	assetID, err := githubAssetRelease.getAssetID(assets)
+	if err != nil {
+		return err
+	}
+
+	// download plugin asset zip file and save it as ReadCloser
+	rc, err := githubAssetRelease.getPluginAssetAsReadCloser(assetID)
+	if err != nil {
+		return err
+	}
```

Comment:
should we have a `defer rc.Close()` here?

Reply:
done